source/common/matcher/list_matcher.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "envoy/matcher/matcher.h"
+
+#include "common/matcher/field_matcher.h"
+
+namespace Envoy {
+namespace Matcher {
+
+/**
+ * A match tree that iterates over a list of matchers to find the first one that matches. If one
+ * does, the MatchResult will be the one specified by the individual matcher.
+ */
+template <class DataType> class ListMatcher : public MatchTree<DataType> {
+public:
+  explicit ListMatcher(absl::optional<OnMatch<DataType>> on_no_match) : on_no_match_(on_no_match) {}
+
+  typename MatchTree<DataType>::MatchResult match(const DataType& matching_data) override {
+    for (const auto& matcher : matchers_) {
+      const auto maybe_match = matcher.first->match(matching_data);
+
+      // One of the matchers don't have enough information, bail on evaluating the match.
+      if (!maybe_match) {
+        return {false, {}};
+      }
```

Comment:
At first I thought this was similar to my other comment, but in thinking about it more I think this is correct. Can you add more comments about why we delay the FIFO match if we hit an entry that doesn't have enough data?